csharp/src/Ice/OutputStream.cs

```diff
@@ -1305,42 +1252,47 @@ private void Expand(int n)
-        /// according to the length in bytes of the numeric sequence.</summary>
-        /// <param name="array">The numeric array to write to the stream.</param>
-        private void WriteNumericSeq(Array array)
-        {
-            int length = Buffer.ByteLength(array);
-            Expand(length);
-            Size += length;
-            int offset = _tail.Offset;
-            int remaining = Math.Min(_currentSegment.Count - offset, length);
-            if (remaining > 0)
-            {
-                Debug.Assert(_currentSegment.Array != null);
-                Buffer.BlockCopy(array, 0, _currentSegment.Array, _currentSegment.Offset + offset, remaining);
-                _tail.Offset += remaining;
-                length -= remaining;
-            }
-
-            if (length > 0)
-            {
-                _currentSegment = _segmentList[++_tail.Segment];
-                Debug.Assert(_currentSegment.Array != null);
-                Buffer.BlockCopy(array, remaining, _currentSegment.Array, _currentSegment.Offset, length);
-                _tail.Offset = length;
-            }
-        }
-
-        /// <summary>Helper method used to write numeric types to the stream.</summary>
+        /// <summary>Writes a fixed-size numeric to the stream.</summary>
         /// <param name="v">The numeric value to write to the stream.</param>
         /// <param name="elementSize">The size in bytes of the numeric type.</param>
-        private void WriteNumeric<T>(T v, int elementSize) where T : struct
+        private void WriteFixedSizeNumeric<T>(T v, int elementSize) where T : struct
         {
+            Debug.Assert(elementSize > 1); // for size 1, we write the byte directly
             Span<byte> data = stackalloc byte[elementSize];
             MemoryMarshal.Write(data, ref v);
-            WriteSpan(data);
+            WriteByteSpan(data);
+        }
+
+        /// <summary>Writes a sequence of fixed-size numeric type, such as int and long, to the stream.</summary>
+        /// <param name="v">The sequence of numeric types.</param>
+        // This method works because (as long as) there is no padding in the memory representation of the
+        // ReadOnlySpan.
+        private void WriteFixedSizeNumericSeq<T>(ReadOnlySpan<T> v) where T : struct
+        {
+            WriteSize(v.Length);
+            WriteByteSpan(MemoryMarshal.AsBytes(v));
+        }
+
+        /// <summary>Writes a tagged sequence of fixed-size numeric type to the stream.</summary>
+        /// <param name="tag">The tag.</param>
+        /// <param name="v">The sequence to write.</param>
+        /// <param name="elementSize">The fixed size of each element of the sequence.</param>
+        private void WriteFixedSizeNumericSeq<T>(int tag, ReadOnlySpan<T> v, int elementSize) where T : struct
+        {
+            Debug.Assert(elementSize > 0);
+
+            // A null T[]? or List<T>? is implicitly converted into a default aka null ReadOnlyMemory<T> or
```

Comment:
Wasn't the default for ReadOnlyMemory empty, and that is why we have to use ReadOnlySpan?

Reply:
The default for both `ReadOnlyMemory` and `ReadOnlySpan` are empty (Length == 0), and yet they are distinguishable from a real empty container "source".

We have to use `ReadOnlyMemory` for the mapping because a ref struct cannot be a type-parameter of something else, like a ValueTask or an OutputStreamWriter.